### DISCUSSION TITLE
implementation of authentication in flutter

### DISCUSSION THREAD

**Discussion by nikhil-varma-mudunuri** [pocketbase/pocketbase#2924] [Q&A] [answered]
I cannot access auth state changes in Flutter while using (onChange). can you share the code implementation for it.

**Comment by ganigeorgiev** [comment]
The `onChange` stream is triggered **only** when there are direct changes to the `pb.authStore` (eg. `save()` or `clear()` was called).

If you want to listen to user changes, you can call `subscribe()`, eg.:
```dart
// subscribe to changes if the current authenticated model
pb.collection("users").subscribe(pb.authStore.model.id, (e) { ... })
```

If the above doesn't help, please provide a code sample and a little more context about use case.

**Comment by nikhil-varma-mudunuri** [comment]
I want to implement user authentication with a pocket base and store the state of the user in the provider so I can access them throughout the app, 
` Stream<AuthStoreEvent> get authStateChanges => admin.pb.authStore.onChange;`
`StreamBuilder(
        stream: Auth().authStateChanges,
        builder: (context, snapshot) {
          print(snapshot.data);
          return Container(
            child: Text("ok"),
          );
        });`
in the above code, I am not receiving any data when the user state changes (like login and logout). 

and another thing does Dart - SDK stores auth data locally or do we need to store them from our side?

**Reply by ganigeorgiev** [reply, marked as answer]
I don't understand the above code sample and it doesn't show how the PocketBase client was initialized and when the authentication is performed. Here is a minimal example that should work (with both admins and auth collections):

```dart
final pb = PocketBase("http://127.0.0.1:8090");

pb.authStore.onChange.listen((e) {
  print("Changed $e");
});

pb.admins.authWithPassword("test@example.com", "1234567890");
```

> and another thing does Dart - SDK stores auth data locally or do we need to store them from our side?

The Dart SDK AuthStore is not persistent at the moment (see https://github.com/pocketbase/dart-sdk#authstore). As a very rudimentary example for a custom store you can check https://github.com/pocketbase/pocketbase/discussions/1887#discussioncomment-5057297.

**Reply by palash018** [reply]
I'm doing it similarly but circular progress indicator is being shown. Is it because the stream will trigger only when there's a change in authstore 
``` 
home:StreamBuilder(stream:AuthMethods.pb.authStore.onChange , builder:(content,snapshot){
          print(snapshot);
          if(snapshot.connectionState==ConnectionState.active){
            if(snapshot.hasData){
              return const Center(child:Text("Successfully logged in"));
            }
            else if(snapshot.hasError){
              return Center(child:Text(snapshot.error.toString()));
            }
          }
          if(snapshot.connectionState==ConnectionState.waiting){
            return const Center(child: CircularProgressIndicator(
              color: primaryColor,
            ));
          }
        return const LoginScreen();
      })
      ```
